source/core/DualTypeEqual.cpp

```diff
@@ -0,0 +1,213 @@
+                UInt8 uInt8ValueY = *static_cast<UInt8*>(pDataY);
+                areEqual = (uInt8ValueX == uInt8ValueY);
+                break;
+            }
+            case 2: {
+                UInt16 uInt16ValueX = *static_cast<UInt16*>(pDataX);
+                UInt16 uInt16ValueY = *static_cast<UInt16*>(pDataY);
+                areEqual = (uInt16ValueX == uInt16ValueY);
+                break;
+            }
+            case 4: {
+                UInt32 uInt32ValueX = *static_cast<UInt32*>(pDataX);
+                UInt32 uInt32ValueY = *static_cast<UInt32*>(pDataY);
+                areEqual = (uInt32ValueX == uInt32ValueY);
+                break;
+            }
+            case 8: {
+                UInt64 uInt64ValueX = *static_cast<UInt64*>(pDataX);
+                UInt64 uInt64ValueY = *static_cast<UInt64*>(pDataY);
+                areEqual = (uInt64ValueX == uInt64ValueY);
+                break;
+            }
+        }
+        return areEqual;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeEqual::ApplyS2CInts(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY)
+    {
+        bool areEqual = false;
+        switch (typeRefX->TopAQSize()) {
+            case 0:
+                areEqual = true;
+                break;
+            case 1: {
+                Int8 int8ValueX = *static_cast<Int8*>(pDataX);
+                Int8 int8ValueY = *static_cast<Int8*>(pDataY);
+                areEqual = (int8ValueX == int8ValueY);
+                break;
+            }
+            case 2: {
+                Int16 int16ValueX = *static_cast<Int16*>(pDataX);
+                Int16 int16ValueY = *static_cast<Int16*>(pDataY);
+                areEqual = (int16ValueX == int16ValueY);
+                break;
+            }
+            case 4: {
+                Int32 int32ValueX = *static_cast<Int32*>(pDataX);
+                Int32 int32ValueY = *static_cast<Int32*>(pDataY);
+                areEqual = (int32ValueX == int32ValueY);
+                // gPlatform.IO.Printf("are %d and %d equal? %s\n", int32ValueX, int32ValueY, areEqual ? "true" : "false");
+                break;
+            }
+            case 8: {
+                Int64 int64ValueX = *static_cast<Int64*>(pDataX);
+                Int64 int64ValueY = *static_cast<Int64*>(pDataY);
+                areEqual = (int64ValueX == int64ValueY);
+                break;
+            }
+        }
+        return areEqual;
+    }
+
+    //------------------------------------------------------------
+    bool DualTypeEqual::ApplyIEEE754Binaries(TypeRef typeRefX, void* pDataX, TypeRef typeRefY, void* pDataY)
+    {
+        bool areEqual = false;
+        if (typeRefX->TopAQSize() == sizeof(Single)) {
+            Single singleValueX = *static_cast<Single*>(pDataX);
+            Single singleValueY = *static_cast<Single*>(pDataY);
+            areEqual = (singleValueX == singleValueY);
```

Comment:
Use something like this:
`areEqual = (singleValueX == singleValueY || (singleValueX != singleValueX && singleValueY != singleValueY))`
to easily check for both values being NaN (which should be considered equal in a variant comparison).

Reply:
Using `::isnan` instead for better readability. It is also being used all over our codebase and should be easier to maintain.